### PR TITLE
add finetune tutorial

### PR DESCRIPTION
**Pull request by yuedongli1**
Thank you for your contribution to the MindYOLO repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindyolo/blob/master/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: tutorials/custom_dataset_finetune.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# 自定义数据集finetune入门
```
Comment (zhanghuiyao) [inline tutorials/custom_dataset_finetune.md:1]: file_name -> custom_dataset.md

File: mindyolo/utils/trainer_factory.py [hunk 2]
```diff
@@ -224,16 +224,7 @@ def train_with_datasink(
...
-        def modify_dataset_columns(image, labels, img_files):
-            return image, labels
-
-        loader = self.dataloader.map(
-            modify_dataset_columns,
-            input_columns=["image", "labels", "img_files"],
-            output_columns=["image", "labels"],
-            column_order=["image", "labels"],
-        )
-
```
Comment (zhanghuiyao) [inline mindyolo/utils/trainer_factory.py:236]: 本处已在2.0适配pr中修复

File: mindyolo/utils/shwd2yolo.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
+import os
```
Comment (zhanghuiyao) [inline mindyolo/utils/shwd2yolo.py:1]: 这个放到外层mindyolo/examples/finetune_SHWD/shwd2yolo.py

File: tutorials/custom_dataset_finetune.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                │          └── 00006564.jpg
+                └── labels
+                      └── train
+                            ├── 00000001.txt
+                            └── 00000002.txt
+```
+
+### 举例
+
+下面以安全帽佩戴检测数据集(SHWD)为例，介绍自定义数据集在MindYOLO上进行finetune的主要流程。
```
Comment (zhanghuiyao) [inline tutorials/custom_dataset_finetune.md:82]: 链接一个 examples/finetune_SHWD/README.md

File: examples/finetune_SHWD/shwd.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  train_set: ./SHWD/train.txt 
+  val_set: ./SHWD/val.txt
+
+  nc: 2
+
+  # class names
+  names: [ 'person',  'hat' ]
+
+  train_transforms: []
+  test_transforms: []
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/shwd.yaml:13]: 1. 是否可以直接继承于外层的./configs/yolov7/yolov7-tiny.yaml，然后在shwd.yaml中修改变化的参数，本目录下的"hyp/yolov7-tiny.yaml"删除掉
2. 末尾增加空行

File: examples/finetune_SHWD/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+000022
+000027
+000034
+```
+
+由于MindYOLO在验证阶段选用图片名称作为image_id，因此图片名称只能为数值类型，而不能为字符串类型，还需要对图片进行改名。对SHWD数据集格式的转换包含如下步骤，详细实现可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/mindyolo/examples/finetune_SHWD/convert_shwd2yolo.py)。
+* 将图片复制到相应的路径下并改名
+* 在根目录下相应的txt文件中写入该图片的相对路径
+* 解析xml文件，在相应路径下生成对应的txt标注文件
+* 验证集还需生成最终的json文件
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:37]: 1. link -> convert_shwd2yolo.py
2. 把转换的流程写清楚，包括怎么运行convert_shwd2yolo.py得到什么形式的数据集

File: examples/finetune_SHWD/yolov7-tiny.yaml [hunk 7]
```diff
@@ -0,0 +1,123 @@
+__BASE__: [
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/yolov7-tiny.yaml:1]: file name -> yolov7-tiny_shwd.yaml

File: examples/finetune_SHWD/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+000034
+```
+
+由于MindYOLO在验证阶段选用图片名称作为image_id，因此图片名称只能为数值类型，而不能为字符串类型，还需要对图片进行改名。对SHWD数据集格式的转换包含如下步骤：
+* 将图片复制到相应的路径下并改名
+* 在根目录下相应的txt文件中写入该图片的相对路径
+* 解析xml文件，在相应路径下生成对应的txt标注文件
+* 验证集还需生成最终的json文件
+
+详细实现可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/mindyolo/examples/finetune_SHWD/convert_shwd2yolo.py)。运行方式如下：
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:39]: link -> convert_shwd2yolo.py

File: examples/finetune_SHWD/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  ```shell
+  cd mindyolo
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:42]: cd mindyolo去掉，默认在mindyolo工程目录下

File: examples/finetune_SHWD/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  python examples/finetune_SHWD/convert_shwd2yolo.py --root_dir ROOT_DIR
+  ```
+
+其中ROOT_DIR为原SHWD数据集所在路径，运行以上命令将在不改变原数据集的前提下，在同级目录生成yolo格式的SHWD数据集。
+
+#### 预训练模型文件转换
+
+由于SHWD数据集只有7000+张图片，选择yolov7-tiny进行该数据集的训练，可下载MindYOLO提供的在coco数据集上训练好的[模型文件](https://github.com/mindspore-lab/mindyolo/blob/master/MODEL_ZOO.md)作为预训练模型。由于coco数据集含有80种物体类别，SHWD数据集只有两类，需将预训练模型文件的最后一层head层去掉， 可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/convert_yolov7_headless.py)。
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:50]: 1. link -> convert_yolov7_headless.py
2. 给个运行命令
3. 把head更nc有关加到说明中

File: examples/finetune_SHWD/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+* 将图片复制到相应的路径下并改名
+* 在根目录下相应的txt文件中写入该图片的相对路径
+* 解析xml文件，在相应路径下生成对应的txt标注文件
+* 验证集还需生成最终的json文件
+
+详细实现可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/mindyolo/examples/finetune_SHWD/convert_shwd2yolo.py)。运行方式如下：
+
+  ```shell
+  cd mindyolo
+  python examples/finetune_SHWD/convert_shwd2yolo.py --root_dir ROOT_DIR
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:43]: ROOT_DIR 换成具体的路径，如 /path_to_hswd/HSWD

File: examples/finetune_SHWD/README.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+由于SHWD数据集只有7000+张图片，选择yolov7-tiny进行该数据集的训练，可下载MindYOLO提供的在coco数据集上训练好的[模型文件](https://github.com/mindspore-lab/mindyolo/blob/master/MODEL_ZOO.md)作为预训练模型。由于coco数据集含有80种物体类别，SHWD数据集只有两类，需将预训练模型文件的最后一层head层去掉， 可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/convert_yolov7_headless.py)。
+
+#### 模型训练
+
+简要的训练流程可参考[finetune_shwd.py](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/finetune_shwd.py)
+
+* 在多卡NPU/GPU上进行分布式模型训练，以8卡为例:
+
+  ```shell
+  cd mindyolo
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:59]: delete "cd mindyolo"

File: examples/finetune_SHWD/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  mpirun --allow-run-as-root -n 8 python examples/finetune_SHWD/finetune_shwd.py --config ./examples/finetune_SHWD/yolov7-tiny.yaml --is_parallel True
+  ```
+
+* 在单卡NPU/GPU/CPU上训练模型：
+
+  ```shell
+  cd mindyolo
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:66]: delete "cd mindyolo"

File: examples/finetune_SHWD/README.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  python examples/finetune_SHWD/convert_shwd2yolo.py --root_dir ROOT_DIR
+  ```
+
+其中ROOT_DIR为原SHWD数据集所在路径，运行以上命令将在不改变原数据集的前提下，在同级目录生成yolo格式的SHWD数据集。
+
+#### 预训练模型文件转换
+
+由于SHWD数据集只有7000+张图片，选择yolov7-tiny进行该数据集的训练，可下载MindYOLO提供的在coco数据集上训练好的[模型文件](https://github.com/mindspore-lab/mindyolo/blob/master/MODEL_ZOO.md)作为预训练模型。由于coco数据集含有80种物体类别，SHWD数据集只有两类，需将预训练模型文件的最后一层head层去掉， 可参考[代码](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/convert_yolov7_headless.py)。
+
+#### 模型训练
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:52]: 模型训练 -> 模型微调(Finetune) 这样会不会好一点

File: examples/finetune_SHWD/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  ```
+
+* 在单卡NPU/GPU/CPU上训练模型：
+
+  ```shell
+  cd mindyolo
+  python examples/finetune_SHWD/finetune_shwd.py --config ./examples/finetune_SHWD/yolov7-tiny_shwd.yaml 
+  ```
+
+MindYOLO所提供的默认参数下完成yolov7-tiny在SHWD数据集上的训练，即可达到ap50为87.0的精度结果；将lr_init参数由0.01改为0.001，即可实现ap50为89.2的精度结果，高于SHWD官方仓库提供的最高ap50精度88.5。
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:70]: 1. 开头加个 Note
2. “MindYOLO所提供的默认参数下完成yolov7-tiny在SHWD数据集上的训练，即可xxx”
改为
“直接用yolov7-tiny默认coco参数在SHWD数据集上训练，可取得AP50 87.0的精度”
是否更好

File: examples/finetune_SHWD/convert_yolov7_headless.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import mindspore as ms
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/convert_yolov7_headless.py:1]: file name -> conver_yolov7-tiny_pretrain_ckpt.py 这样是否更清晰

File: tutorials/custom_dataset.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
+# 数据集格式介绍
```
Comment (zhanghuiyao) [inline tutorials/custom_dataset.md:1]: 只给一个yolo格式的整体目录说明就可以了，然后下面说自定义数据集转成yolo格式进行训练，example参考hswd

File: examples/finetune_SHWD/README.md [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+000034
+```
+
+由于MindYOLO在验证阶段选用图片名称作为image_id，因此图片名称只能为数值类型，而不能为字符串类型，还需要对图片进行改名。对SHWD数据集格式的转换包含如下步骤：
+* 将图片复制到相应的路径下并改名
+* 在根目录下相应的txt文件中写入该图片的相对路径
+* 解析xml文件，在相应路径下生成对应的txt标注文件
+* 验证集还需生成最终的json文件
+
+详细实现可参考[convert_shwd2yolo.py](https://github.com/mindspore-lab/mindyolo/blob/master/mindyolo/examples/finetune_SHWD/convert_shwd2yolo.py)。运行方式如下：
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:39]: link没改

File: examples/finetune_SHWD/README.md [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  mpirun --allow-run-as-root -n 8 python examples/finetune_SHWD/finetune_shwd.py --config ./examples/finetune_SHWD/yolov7-tiny.yaml --is_parallel True
+  ```
+
+* 在单卡NPU/GPU/CPU上训练模型：
+
+  ```shell
+  python examples/finetune_SHWD/finetune_shwd.py --config ./examples/finetune_SHWD/yolov7-tiny_shwd.yaml 
+  ```
+
+*注意：直接用yolov7-tiny默认coco参数在SHWD数据集上训练，可取得AP50 87.0的精度。将lr_init参数由0.01改为0.001，即可实现ap50为89.2的精度结果，高于SHWD官方仓库提供的最高ap50精度88.5。*
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:71]: 后面这句"高于xxx"不要说；

File: examples/finetune_SHWD/README.md [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+由于SHWD数据集只有7000+张图片，选择yolov7-tiny进行该数据集的训练，可下载MindYOLO提供的在coco数据集上训练好的[模型文件](https://github.com/mindspore-lab/mindyolo/blob/master/MODEL_ZOO.md)作为预训练模型。由于coco数据集含有80种物体类别，SHWD数据集只有两类，模型的最后一层head层输出与类别数nc有关，因此需将预训练模型文件的最后一层去掉， 可参考[convert_yolov7-tiny_pretrain_ckpt.py](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/convert_yolov7-tiny_pretrain_ckpt.py)。运行方式如下：
+
+  ```shell
+  python examples/finetune_SHWD/convert_yolov7-tiny_pretrain_ckpt.py
+  ```
+
+#### 模型微调(Finetune)
+
+简要的训练流程可参考[finetune_shwd.py](https://github.com/mindspore-lab/mindyolo/blob/master/examples/finetune_SHWD/finetune_shwd.py)
```
Comment (zhanghuiyao) [inline examples/finetune_SHWD/README.md:57]: link没改

File: tutorials/custom_dataset.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+                │     │     ├── 00000001.jpg
+                │     │     └── 00000002.jpg
+                │     └── val
+                │          ├── 00006563.jpg
+                │          └── 00006564.jpg
+                └── labels
+                      └── train
+                            ├── 00000001.txt
+                            └── 00000002.txt
+```
```
Comment (zhanghuiyao) [inline tutorials/custom_dataset.md:21]: 后面可以按照这个目录结合具体的文件说明下每个文件的标注格式和含义；

**Review by zhanghuiyao — APPROVED** [review]

**Review by zhanghuiyao — DISMISSED** [review]

File: examples/finetune_SHWD/convert_shwd2yolo.py [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+            image_index += 1
+
+    for categoryname in category_set:
+        addCatItem(categoryname)
+
+    train_txt_yolo.close()
+    val_txt_yolo.close()
+
+    json_file = os.path.join(new_dir, 'annotations', 'instances_val2017.json')
+    json.dump(coco, open(json_file, 'w'))
```
Comment (CaitinZhao) [inline examples/finetune_SHWD/convert_shwd2yolo.py:244]: 教程中没有对关键代码的解释，需要在代码中关键部分加注释

File: examples/finetune_SHWD/convert_yolov7-tiny_pretrain_ckpt.py [hunk 23]
```diff
@@ -0,0 +1,15 @@
+import mindspore as ms
+
+
+def convert_weight(ori_weight, new_weight):
+    new_ckpt = []
+    param_dict = ms.load_checkpoint(ori_weight)
+    for k, v in param_dict.items():
+        if '77' in k:
```
Comment (CaitinZhao) [inline examples/finetune_SHWD/convert_yolov7-tiny_pretrain_ckpt.py:8]: 77为什么要drop掉？

File: examples/finetune_SHWD/finetune_shwd.py [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+        train_url=args.train_url,
+        run_eval=args.run_eval,
+    )
+    logger.info("Training completed.")
+
+
+if __name__ == "__main__":
+    parser = get_parser_train()
+    args = parse_args(parser)
+    train_shwd(args)
```
Comment (CaitinZhao) [inline examples/finetune_SHWD/finetune_shwd.py:149]: 为什么要单加一个文件，和训练的文件很像，有点多余

**Review by CaitinZhao — APPROVED** [review]

**Review by zhanghuiyao — APPROVED** [review]

**Pull request merged**